### PR TITLE
add weight_dequantize python api

### PR DESCRIPTION
**Pull request by yuanlehome**
### PR types

New features
### PR changes

APIs
### Description

- add weight_dequantize python api
- related pr https://github.com/PaddlePaddle/Paddle/pull/57685
### Others
Pcard-71500

### CODE REVIEW

File: paddle/phi/kernels/gpu/weight_dequantize_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                            DataType out_dtype,
+                            DenseTensor* out) {
+#if defined(PADDLE_WITH_CUTLASS)
+  auto out_dims = out->dims();
+  dev_ctx.template Alloc<T>(out);
+  WeightDequantize<T, Context>(dev_ctx, x, scale, algo, true, out);
+  out->Resize({{out_dims[1], out_dims[0]}});
+  auto out_tmp = Transpose<T, Context>(dev_ctx, *out, {1, 0});
+  out->ShareDataWith(out_tmp);
+#endif
```
Comment (MARD1NO) [inline paddle/phi/kernels/gpu/weight_dequantize_kernel.cu:43]:
```cpp
# else
PADDLE_THROW_ERROR(xxxx)
```

**Review by MARD1NO — APPROVED** [review]

File: python/paddle/nn/quant/quantized_linear.py [hunk 2]
```diff
@@ -60,6 +59,53 @@ def weight_quantize(x, algo="weight_only_int8"):
...
+
+            import paddle
+            from paddle.nn.quant import weight_quantize, weight_dequantize
+
+            paddle.seed(2023)
+            x = paddle.rand(shape=[64, 32], dtype=paddle.float16)
+            out, scale = weight_quantize(x, algo='weight_only_int8')
+
+            x_dequant = weight_dequantize(out, scale)
+            print(paddle.sum(paddle.mean(paddle.abs(x-x_dequant))))
```
Comment (SigureMo) [inline python/paddle/nn/quant/quantized_linear.py]:
```suggestion
            >>> import paddle
            >>> from paddle.nn.quant import weight_quantize, weight_dequantize

            >>> paddle.seed(2023)
            >>> x = paddle.rand(shape=[64, 32], dtype=paddle.float16)
            >>> out, scale = weight_quantize(x, algo='weight_only_int8')

            >>> x_dequant = weight_dequantize(out, scale)
            >>> print(paddle.sum(paddle.mean(paddle.abs(x-x_dequant))))
```

需要使用新的写法，`weight_quantize` 也需要修改

Comment (yuanlehome) [reply python/paddle/nn/quant/quantized_linear.py]: done.

**Review by sunzhongkai588 — APPROVED** [review]
LGTM

**Review by zyfncg — APPROVED** [review]

**Comment by yuanlehome** [comment]
> LGTM

TODO:
- 发现这几个weight quantize 相关 api都没有在官网提供文档，后面统一加下

**Review by Aurelius84 — APPROVED** [review]

**Review by zhangbo9674 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**